agent/src/beerocks/slave/agent_db.h

```diff
@@ -0,0 +1,76 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2019-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include "task_pool.h"
+
+#include <iostream>
+#include <memory>
+#include <mutex>
+
+using namespace beerocks_message;
+using namespace son;
+using namespace beerocks;
+
+/**
+ * @brief AgentDB is a class used by all main Agent threads on the Agent process to store the Agent
+ * common data.
+ * The class is implemented using singleton design pattern.
+ * It is thread safe and being locked and released automatically.
+ * 
+ * How to use:
+ * 
+ * auto db = AgentDB::get();            // Get DB and automatically lock it. The lock will be
+ *                                      // Release when 'db' will be destructed.
+ * 
+ * db->foo = 42;                        // Change database member 'foo' to value 42
+ * db->bar = std::string("prpl")        // Set another database member
+ * 
+ * AgentDB::get()->foo = 44;            // Set database member directly.
+ * auto my_foo = AgentDB::get()->foo;   // Get a copy of database member.
+ * 
+ * auto &foo = AgentDB::get()->foo;     // Unsafe! Don't do it! The database will be unlocked,
+ *                                      // but the reference will remain.
```

Comment:
This is getting me worried a little...

A typical use case would be something like this:
```
for (foo : AgentDB::get()->foo_list) ...
```
This doesn't work as expected, because the lock is only held during the evaluation of the loop expression, not during the body of the loop. You *have* to write it like this:
```
auto db = AgentDB::get();
for (foo : db->foo_list) ...
```

Therefore, I think it's better to disallow any other use than `auto db = AgentDB::get();`. That makes it *much* easier to check that the code is correct.

(Note: lifetimes of temporaries are extremely difficult to reason about, so my analysis may be incorrect. But, since they are extremely difficult to reason about, there's no way to be sure if there are no situations that the lifetime is too short. So better safe than sorry...)

Therefore, I would write:
```suggestion
 * AgentDB::get()->foo = 44;            // Set database member directly.
 * auto my_foo = AgentDB::get()->foo;   // Unsafe! Don't do it! 
 *                                      // This specific example is OK, but it's hard to know when the database will be unlocked.
 * auto &foo = AgentDB::get()->foo;     // Unsafe! Don't do it! The database will be unlocked,
 *                                      // but the reference will remain.
```

Another tricky situation is where you get a nested reference from the database - that's actually the same as a reference, but less obvious. For example, let's say the database has a member `std::shared_ptr<TopologyTask> topology_task`. Then `auto task = db->topology_task;` will get you a reference (implied through the smart pointer). So we always have to be careful about when the db is going to be unlocked.

Bottom line: the SafeDB is a useful tool, but it's not sufficient by far to be safe.

We could make it more reliable by saying that the database must not contain any references to begin with. So that would already exclude task_pool, which has references to tasks... It means that references should be managed through identifiers, and every time you want to dereference something, you have to go through the database. But even that is not a silver bullet, because when you start passing those identifiers around, you can end up with unlocking the database in-between and consistency is no longer guaranteed (e.g. you have some function that takes a station identifier out of the database and returns it; if the caller hasn't locked the database already, the database is unlocked when the function returns, and by the time the caller locks the database again to look up the reference, it is already gone).

Bottom line (really, this time): I prefer not to have to deal with multithreaded code :-)

Reply:
I added/changed the "how to use".
I think we should give it a shot.